MapboxNavigation/NavigationMapView.swift

```diff
@@ -671,6 +676,188 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
     }
 
     // MARK: - Vanishing route line methods
+
+    private var primaryRoutePoints: RoutePoints?
+    private var primaryRouteLineGranularDistances: RouteLineGranularDistances?
+    private var primaryRouteRemainingDistancesIndex: Int?
+    public var newFractionTraveled: Double = 0.0
+    
+    private struct RoutePoints {
+        var nestedList: [[[CLLocationCoordinate2D]]]
+        var flatList: [CLLocationCoordinate2D]
+    }
+    
+    private struct RouteLineGranularDistances {
+        var distance: Double
+        var distanceArray: [RouteLineDistancesIndex]
+    }
+    
+    private struct RouteLineDistancesIndex {
+        var point: CLLocationCoordinate2D
+        var distanceRemaining: Double
+    }
+    
+    private func initPrimaryRoutePoints(route: Route) {
+        primaryRoutePoints = parseRoutePoints(route: route)
+        primaryRouteLineGranularDistances = calculateRouteGranularDistances(coordinates: primaryRoutePoints?.flatList ?? [])
+    }
+    
+    /**
+    Tranform the route data into nested arrays of legs -> steps -> coordinates.
+    The first and last point of adjacent steps overlap and are duplicated.
+    */
+    private func parseRoutePoints(route: Route) -> RoutePoints {
```

Comment:
What about reusing already implemented functionality which allows to create list of coordinates along the route? E.g. there is code in:
https://github.com/mapbox/mapbox-navigation-ios/blob/8bb50396f6715f36b29cc729915ee0a65404eec4/MapboxNavigation/NavigationMapView.swift#L1148-L1154 which partly does this job. We can extract that part into method and reuse it. Also as in previous comment, do we really need to create `RoutePoints`? What about just returning list of coordinates?